ath79: add support for Teltonika RUT230

arturaspaulius opened the pull request:
Commit adds support for teltonika rut2xx routers.

Specification:
- 400 MHz CPU
- 64 MB of RAM (DDR2)
- 16 MB of FLASH (SPI NOR)
- 2x RJ45 10/100 Mbps Ethernet ports
- 2T2R 2,4 GHz (AR9330)
- built-in 4G/3G module Quectel UC20-E
- 11x LED
- 1 x SMA for 3G,  1x RP-SMA for WiFi antenna connectors
- 1x button (reset)
- DC jack for main power input (9-30 V)

Flash instruction:

Vendor firmware is based on OpenWrt CC release. Use the "factory" image
directly in GUI (make sure to uncheck "keep settings") or in U-Boot web
based recovery.

Signed-off-by: Artūras Paulius <arturas.paulius@gmail.com>

adrianschmutzler commented:
The commit description and Signed-off-by should go into the commit message.

For the commit title, please add "ath79:" prefix and use imperative form "ath79: add support for ...". Despite, in the code you are effectively adding _one_ device (rut230), so please also tell that in the commit title.

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+		bootargs = "console=ttyATH0,115200 root=31:05 rootfstype=squashfs";
+	};
+
+
```
Don't add double empty lines.

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+		serial0 = &uart;
+		led-boot = &signal3;
+		led-failsafe = &signal3;
+		led-upgrade = &signal3;
```
Why that LED? Does this mirror vendor behavior?

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+
+		sim {
+			label = "SIM holder";
+			linux,code = <KEY_RFKILL>;
```
This does not match the name/label. Is this vendor behavior?

arturaspaulius replied:
Yes, not sure what for, but it was hard coded in the vendor firmware

adrianschmutzler replied:
Still, if it's called "SIM holder", it shouldn't do an RFKILL. So maybe it's better to replace the linux,code by BTN_1.
Despite, I do not find input/sim in the documentation:
https://wiki.teltonika-networks.com/view/RUT230_front_and_back_panels

arturaspaulius replied:
for sim its triggered to gpio when sim cart is inserted and apparently it reloads modem without rebooting it and for the input, it is in the power socket. 
https://wiki.teltonika-networks.com/view/RUT230_Input/Output

adrianschmutzler replied:
However, RFKILL won't reboot, but just toggle on/off ...?

arturaspaulius replied:
Ok, i found something. rfkill is defined in /etc/rc.button/ on the vendor firmware:
#!/bin/sh

. /lib/functions.sh
modem=$(gsmctl -n -y | head -c 4)

logger "$BUTTON was $ACTION for $SEEN seconds"

if [ "${ACTION}" = "released" ]
then
	if [ "$modem" = "EC25" -o "$modem" = "UC20" ]
	then
		ifdown ppp
		sleep 2
		/etc/init.d/operctl stop
		gsmctl -A "AT+CFUN=0"
	else
		/etc/init.d/modem stop
	fi

	/etc/init.d/ledsman restart
elif [ "${ACTION}" = "pressed" ]
then
	if [ "$modem" = "EC25" -o "$modem" = "UC20" ]
	then
		gsmctl -A "AT+CFUN=1"
		sleep 1
		/etc/init.d/gsmd restart
	else
		sleep 1
		/etc/init.d/modem restart
	fi
fi 

So its just restarts some services and changes modem functionality. This just waits for state to change but the gpio button itself toggles on/off when inserting/removing sim card. I guess it won't hurt to change linux code itself then, because same thing can be added for BTN_1 in rc.button ?

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+
+		signal3: signal3 {
```
Please include "led_" prefix for the DT label (as this is used in a global context):

`led_signal3: signal3 {`

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
```
Add empty line after status.

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+				read-only;
+			};
+
+			art: partition@30000 {
```
Pleaso only add DT labels ("art:") where needed (should be config and art...)

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+	status = "okay";
+
+	mtd-mac-address = <&config 0x0>;
+	mtd-mac-address-increment = <1>;
```
Is there only one address on flash? Does the assignment match vendor firmware?

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+&gpio {
+	status = "okay";
+
+	gpio_usb_power {
```
Just "usb_power".

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+};
+
+&wmac {
+	status = "okay";
```
Add empty line after status.

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,200 @@
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&config 0x0>;
+	mtd-mac-address-increment = <2>;
```
Is there a valid address in 0x1002/0x1006, which would be read by caldata?

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -142,6 +142,10 @@ pcs,cr3000)
 	;;
+teltonika,rut230)
+       ucidef_set_led_switch "lan" "LAN" "rut230:green:lan" "eth0"
+       ucidef_set_led_switch "wan" "WAN" "rut230:green:wan" "eth1"
```
This tells me you have not checked LEDs, as this won't work. You are mixing switch and netdev syntax here.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  SOC := ar9330
+  DEVICE_VENDOR := Teltonika
+  DEVICE_MODEL := RUT230
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-chipidea2 kmod-usb-net \
```
kmod-usb-core and kmod-usb-serial can be removed here, as they will be selected by the other packages.

adrianschmutzler replied:
kmod-usb-net can be removed as well.
Despite, wrap lines to not exceed 80 characters.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  DEVICE_MODEL := RUT230
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-chipidea2 kmod-usb-net \
+       kmod-usb-net-qmi-wwan uqmi -uboot-envtools kmod-usb-serial kmod-usb-serial-option comgt usb-modeswitch
+  BOARDNAME := RUT230
```
That shouldn't be required anymore.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  SUPPORTED_DEVICES += rut230 
+endef
+TARGET_DEVICES += teltonika_rut230
+
```
no double empty lines.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  TPLINK_HWID := 0x32200002
+  TPLINK_HWREV := 0x1
+  TPLINK_HEADER_VERSION := 1
+  REVISION := RUT2xx
```
That's only for upgrade from ar71xx, which doesn't contain this device.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  IMAGES := factory.bin sysupgrade.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs |\
+       pad-rootfs | teltonika-fw-fake-checksum | append-string RUT200019999master |\
+       append-md5sum-bin | check-size $$$$(IMAGE_SIZE)
```
check-size can be used without argument now. Same for sysupgrade.bin below.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1123,32 @@ define Device/teltonika_rut955
+  BOARDNAME := RUT230
+  IMAGE_SIZE := 15552k
+  KERNEL := kernel-bin | append-dtb | lzma | tplink-v1-header
+  KERNEL_INITRTH_SOCAMFS := $$(KERNEL)
```
KERNEL_INITRAMFS := $$(KERNEL)
?

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -40,6 +40,10 @@ define Build/append-md5sum-bin
+define Build/append-string
+       echo -n $(1) >> $@
+endef
+
```
That's redundant, is this is also present in include/image-commands.mk.
Look at rut955 in this file, it's already using it.

adrianschmutzler commented:
You don't touch board.d/02_network. That's okay for a two-port if the LAN/WAN arrangement matches vendor firmware. Is this the case?
If it's swapped, you'd need to adjust in 02_network.

arturaspaulius commented:
> You don't touch board.d/02_network. That's okay for a two-port if the LAN/WAN arrangement matches vendor firmware. Is this the case?
> If it's swapped, you'd need to adjust in 02_network.

Yes, there is to two ports. Now leds blink like it should.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1119,6 +1119,30 @@ define Device/teltonika_rut955
+  TPLINK_HWID := 0x32200002
+  TPLINK_HWREV := 0x1
+  TPLINK_HEADER_VERSION := 1
+  SUPPORTED_DEVICES += rut230 
```
As stated already, SUPPORTED_DEVICES is not required here.

adrianschmutzler commented:
Thanks for addressing most of the remarks so far. Please fix (=remove) the indent in the commit message (looks like copy/paste from git log) and fix spaces/capitalization in your Signed-off-by name.

Despite, can you provide an overview of MAC address location in flash (config/art partitions) and check what's the MAC address assignment in vendor firmware?
Maybe you can provide an overview like shown here: https://openwrt.org/docs/guide-developer/mac.address#merge_your_data

arturaspaulius commented:
> Thanks for addressing most of the remarks so far. Please fix (=remove) the indent in the commit message (looks like copy/paste from git log) and fix spaces/capitalization in your Signed-off-by name.
> 
> Despite, can you provide an overview of MAC address location in flash (config/art partitions) and check what's the MAC address assignment in vendor firmware?
> Maybe you can provide an overview like shown here: https://openwrt.org/docs/guide-developer/mac.address#merge_your_data

I found this:
root@Teltonika:~# hexdump -C /dev/mtd1 | grep '00 1e 42'
00000000  00 1e 42 18 eb b1 ff ff  ff ff ff ff ff ff ff ff  |..B.............|
root@Teltonika:~# hexdump -C /dev/mtd2 | grep '00 1e 42'

mtd1 is config partition and mtd2 is art partition.
I found only one location in config partition.

adrianschmutzler commented:
Note that technically your grep wouldn't work if the grepped part is wrapped into the next line. But since the address is followed by ffs, I don't think there is much to find there.

So now, like mentioned earlier, we just need the address assignment in vendor firmware to find the correct offsets. Just flash back and report "ip a" or "ifconfig" here.

arturaspaulius commented:
> Note that technically your grep wouldn't work if the grepped part is wrapped into the next line. But since the address is followed by ffs, I don't think there is much to find there.
> 
> So now, like mentioned earlier, we just need the address assignment in vendor firmware to find the correct offsets. Just flash back and report "ip a" or "ifconfig" here.

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br-lan state UP group default qlen 1000
    link/ether 00:1e:42:18:eb:b1 brd ff:ff:ff:ff:ff:ff
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN group default qlen 1000
    link/ether 00:1e:42:18:eb:b2 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::21e:42ff:fe18:ebb2/64 scope link 
       valid_lft forever preferred_lft forever
4: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN group default 
    link/tunnel6 :: brd ::
5: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default 
    link/sit 0.0.0.0 brd 0.0.0.0
6: ifb0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 32
    link/ether f6:b6:bb:01:e8:3c brd ff:ff:ff:ff:ff:ff
7: ifb1: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 32
    link/ether 7e:ba:0e:36:ff:f8 brd ff:ff:ff:ff:ff:ff
8: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN group default 
    link/gre 0.0.0.0 brd 0.0.0.0
9: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
10: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN group default 
    link/gre6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
11: wwan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether 82:56:f8:55:fb:f0 brd ff:ff:ff:ff:ff:ff
    inet 10.138.89.221/30 brd 10.138.89.223 scope global wwan0
       valid_lft forever preferred_lft forever
    inet6 fe80::8056:f8ff:fe55:fbf0/64 scope link 
       valid_lft forever preferred_lft forever
13: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 00:1e:42:18:eb:b1 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fe80::21e:42ff:fe18:ebb1/64 scope link 
       valid_lft forever preferred_lft forever
14: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP group default 
    link/ether 00:1e:42:18:eb:b3 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::21e:42ff:fe18:ebb3/64 scope link 
       valid_lft forever preferred_lft forever

adrianschmutzler commented:
Okay, so we have learned the following:
```
MAC address assignment

use    stock      source           OpenWrt
lan    eth0       &config 0x0   eth0
wan  eth1        &config 0x0 +1  eth1
2g    wlan0      &config 0x0 +2  wlan0
```

However, I would have expected the eth0/eth1 to be swapped.

Would you verify
- that LAN and WAN are assigned to the same ports for vendor firmware and OpenWrt
- that br-lan on OpenWrt has the MAC address ending with b1

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,207 @@
+	leds {
+		compatible = "gpio-leds";
+
+		signal_bar0 {
```
I'd keep node name and label consistent here. Just call the node "signal1" as well and do so for the following.

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,207 @@
+
+	aliases {
+		serial0 = &uart;
+		led-boot = &led_signal_bar2;
```
Is this vendor behavior or did you choose the LED randomly?

arturaspaulius replied:
I think i was testing something, removed it.

adrianschmutzler replied:
Okay, but now you have no LEDs for signalling boot/failsafe etc. anymore.
I'd recommend to just use the signal5 LED for those.

arturaspaulius commented:
br-lan    Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B2  
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fdc8:23f1:11d3::1/60 Scope:Global
          inet6 addr: fe80::21e:42ff:fe18:ebb2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:391 errors:0 dropped:0 overruns:0 frame:0
          TX packets:229 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:44914 (43.8 KiB)  TX bytes:47252 (46.1 KiB)

eth0      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B2  
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:417 errors:0 dropped:0 overruns:0 frame:0
          TX packets:237 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:52473 (51.2 KiB)  TX bytes:48853 (47.7 KiB)
          Interrupt:5 

eth1      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B1  
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.254.0
          inet6 addr: fe80::21e:42ff:fe18:ebb1/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1034 errors:0 dropped:0 overruns:0 frame:0
          TX packets:451 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:148497 (145.0 KiB)  TX bytes:60707 (59.2 KiB)
          Interrupt:4 

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:52 errors:0 dropped:0 overruns:0 frame:0
          TX packets:52 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:7003 (6.8 KiB)  TX bytes:7003 (6.8 KiB)

It is indeed swapped. Do I edit 02_network or dts?

adrianschmutzler commented:
Would you verify
- that LAN and WAN are assigned to the same ports for vendor firmware and OpenWrt

arturaspaulius commented:
> Would you verify
> 
> * that LAN and WAN are assigned to the same ports for vendor firmware and OpenWrt

Sorry I don't understand. How I can do that then ?

adrianschmutzler commented:
Your device has two ports. One will give an IP address via DHCP when you connect a device to it. That's LAN.
The other port will listen to an external DHCP server, that's WAN.

So, in practice, you will use a computer and connect it to the two ports. If you get an IP from the router and can connect to the device via 192.168.1.1, then it's the LAN port. This will only happen for one of the two ports.
My question is whether this happens for the port that is labelled LAN when you have the OpenWrt from your commit installed.

adrianschmutzler commented:
... or asked simpler: Does LAN come out of the LAN port?

arturaspaulius commented:
Yes, when i connected WAN port to router it gets ip 192.168.0.70:
eth0      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B2  
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:521 errors:0 dropped:0 overruns:0 frame:0
          TX packets:338 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:92389 (90.2 KiB)  TX bytes:50956 (49.7 KiB)
          Interrupt:5 

eth1      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B1  
          inet addr:192.168.0.70  Bcast:192.168.1.255  Mask:255.255.254.0
          inet6 addr: fe80::21e:42ff:fe18:ebb1/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:386 errors:0 dropped:0 overruns:0 frame:0
          TX packets:209 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:76984 (75.1 KiB)  TX bytes:25970 (25.3 KiB)
          Interrupt:4 
Connecting my pc to LAN port gets also ip
eno1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.114  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fda2:d2f8:4d77::99a  prefixlen 128  scopeid 0x0<global>
        inet6 fda2:d2f8:4d77:0:bd5f:11fa:bcaf:c9ca  prefixlen 64  scopeid 0x0<global>
        inet6 fda2:d2f8:4d77:0:496c:4052:c4ef:4901  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::560e:246d:77f:6cdf  prefixlen 64  scopeid 0x20<link>
        ether e4:e7:49:34:d8:5a  txqueuelen 1000  (Ethernet)
        RX packets 3524855  bytes 4543959889 (4.5 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2210722  bytes 558058348 (558.0 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

arturaspaulius commented:
same from the vendor firmware. It gives address 192.168.0.71. Just that the mac is different in the interfaces so router gives new ip.
eth0      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B1  
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:4 errors:0 dropped:0 overruns:0 frame:0
          TX packets:4 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:316 (316.0 B)  TX bytes:280 (280.0 B)
          Interrupt:5 

eth1      Link encap:Ethernet  HWaddr 00:1E:42:18:EB:B2  
          inet addr:192.168.0.71  Bcast:192.168.1.255  Mask:255.255.254.0
          inet6 addr: fe80::21e:42ff:fe18:ebb2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:444 errors:0 dropped:0 overruns:0 frame:0
          TX packets:228 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:76137 (74.3 KiB)  TX bytes:22499 (21.9 KiB)
          Interrupt:4

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,204 @@
+	status = "okay";
+
+	mtd-mac-address = <&config 0x0>;
+	mtd-mac-address-increment = <1>;
```
Okay, so you just need to move the `mtd-mac-address-increment = <1>;` to &eth0 to fix the MAC addresses, nothing else.

adrianschmutzler commented:
Okay, so put this into the commit message:
```
MAC address assignment

use    stock      source           OpenWrt
lan    eth0       &config 0x0   eth0/&eth1
wan  eth1        &config 0x0 +1  eth1/&eth0
2g    wlan0      &config 0x0 +2  wlan0
```

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,204 @@
+			gpios = <&gpio 24 GPIO_ACTIVE_HIGH>;
+		};
+
+		status_4g {
```
I don't find this LED in the manual?

arturaspaulius replied:
tatic struct gpio_led tlt_rut200_leds_gpio[] __initdata = {
  1         {
  2                 .name           = "signal_bar0",
  3                 .gpio           = TLT_RUT200_GPIO_LED_SS1,
  4                 .active_low     = 1,
  5         }, {
  6                 .name           = "signal_bar1",
  7                 .gpio           = TLT_RUT200_GPIO_LED_SS2,
  8                 .active_low     = 1,
  9         }, {
 10                 .name           = "signal_bar2",
 11                 .gpio           = TLT_RUT200_GPIO_LED_SS3,
 12                 .active_low     = 1,
 13         }, {
 14                 .name           = "signal_bar3",
 15                 .gpio           = TLT_RUT200_GPIO_LED_SS4,
 16                 .active_low     = 1,
 17         }, {
 18                 .name           = "signal_bar4",
 19                 .gpio           = TLT_RUT200_GPIO_LED_SS5,
 20                 .active_low     = 1,
 21         }, {
 22                 .name           = "status_2g",
 23                 .gpio           = TLT_RUT200_GPIO_LED_2G,
 24                 .active_low     = 0,
 25         }, {
 26                 .name           = "status_3g",
 27                 .gpio           = TLT_RUT200_GPIO_LED_3G,
 28                 .active_low     = 0,
 29         }, {
 30                 .name           = "status_4g",
 31                 .gpio           = TLT_RUT200_GPIO_LED_4G,
 32                 .active_low     = 0,
 33         }, {
 34                 .name           = "led_lan",
 35                 .gpio           = 13,
 36                 .active_low     = 0,
 37         }, {
 38                 .name           = "led_wan",
 39                 .gpio           = 14,
 40                 .active_low     = 0,
 41         },
 42 };
This was hardcoded into vendors firmware

arturaspaulius replied:
But i think it is added for RUT240 device witch has 4g and the extra LED witch is the only difference from RUT230. I can remove it, but this won't be compatible with 240

arturaspaulius replied:
In addition they use same firmware for both devices

adrianschmutzler replied:
Yes, I understand that you correctly ported this from their code (as it appears to me).

Still, if the two devices have different features, we should give them different images here. So, I'd limit the coded features for the rut230 to those the rut230 actually has.

arturaspaulius replied:
Sure, I agree. Removed it.

arturaspaulius committed:
ath79: add support for Teltonika RUT230

Specification:

400 MHz CPU
64 MB of RAM (DDR2)
16 MB of FLASH (SPI NOR)
2x RJ45 10/100 Mbps Ethernet ports
2T2R 2,4 GHz (AR9330)
built-in 4G/3G module Quectel UC20-E
11x LED
1 x SMA for 3G, 1x RP-SMA for WiFi antenna connectors
1x button (reset)
DC jack for main power input (9-30 V)

MAC address assignment

use    	stock	source		OpenWrt
lan	eth0	&config 0x0   	eth0/&eth1
wan  	eth1	&config 0x0 +1  eth1/&eth0
2g    	wlan0	&config 0x0 +2  wlan0

Flash instruction:

Vendor firmware is based on OpenWrt CC release. Use the "factory" image
directly in GUI (make sure to uncheck "keep settings") or in U-Boot web
based recovery.

Signed-off-by: Artūras Paulius <arturas.paulius@gmail.com>

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,205 @@
+	model = "Teltonika RUT230";
+
+	chosen {
+		bootargs = "console=ttyATH0,115200 root=31:05 rootfstype=squashfs";
```
Do you really need these additional bootargs?
I've not found any device in ath79 setting anything beyond console ...

adrianschmutzler commented on target/linux/ath79/dts/ar9330_teltonika_rut230.dts:
```diff
@@ -0,0 +1,205 @@
+
+&wmac {
+	status = "okay";
+
```
What about caldata?

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -163,6 +163,10 @@ pcs,cr3000)
 	;;
+teltonika,rut230)
+       ucidef_set_led_netdev "lan" "LAN" "rut230:green:lan" "eth0"
+       ucidef_set_led_netdev "wan" "WAN" "rut230:green:wan" "eth1"
```
For proper link-state detection, I suspect that one statement (probably lan) needs to be set up with ucidef_set_led_switch here. I'd give this low priority, though.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1182,6 +1182,29 @@ define Device/teltonika_rut955-h7v3c0
 endef
 TARGET_DEVICES += teltonika_rut955-h7v3c0
 
+define Device/teltonika_rut230
```
Alphabetic sorting ...

adrianschmutzler commented:
This appears to be abandoned, so I'm closing it now. Please ask for reopening if you want to continue working on it (and _before_ pushing an update to your GitHub branch).
